diff --git a/mingw-w64-crt/crt/crtexe.c b/mingw-w64-crt/crt/crtexe.c
index 94bad6aaf..1f001189a 100644
--- a/mingw-w64-crt/crt/crtexe.c
+++ b/mingw-w64-crt/crt/crtexe.c
@@ -315,3 +315,37 @@ int __cdecl atexit (_PVFV func)
 }
 
 char __mingw_module_is_dll = 0;
+
+/* Set stdio to binary mode to workaround CRT's bugged "double translation".
+   1. UCRT randomly eats linefeeds when piping on Windows Vista.
+      Compiling any non-trivial program compiled with '-g3 -pipe'
+      is likely to fail.
+   2. MSVCRT, and UCRT prior to 10.0.14393, cannot write multi-byte
+      characters one-char-by-one-char after set_locale(LC_ALL, "").
+        int main()
+        {
+          setlocale(LC_ALL, "");
+          for (auto ch : "你好")
+            fputc(ch, stdout);
+        }
+      This is exactly how MinGW-w64's printf and toolchain's NLS work.
+ */
+
+#include <fcntl.h>
+#include <io.h>
+#include <stdio.h>
+#include <windows.h>
+
+static UINT saved_console_cp = 0;
+static UINT saved_console_output_cp = 0;
+
+static void __cdecl set_binary_stdio(void)
+{
+  _setmode(_fileno(stdin), _O_BINARY);
+  _setmode(_fileno(stdout), _O_BINARY);
+  _setmode(_fileno(stderr), _O_BINARY);
+}
+
+__auto_type __mingw_lite_set_binary_stdio
+  __attribute__ ((section (".CRT$XCML"), used))
+  = &set_binary_stdio;
