diff --git a/libgfortran/caf/shmem/shared_memory.c b/libgfortran/caf/shmem/shared_memory.c
index afafd50f80b..eb51a18378a 100644
--- a/libgfortran/caf/shmem/shared_memory.c
+++ b/libgfortran/caf/shmem/shared_memory.c
@@ -233,9 +233,14 @@ shared_memory_init (shared_memory_act *mem, size_t size)
       exit (1);
     }
 #elif defined(WIN32)
+#if defined(_WIN32_WINNT) && _WIN32_WINNT < 0x0600
+  mem->glbl.base
+    = (LPTSTR) MapViewOfFileEx (mem->shm_fd, FILE_MAP_ALL_ACCESS, 0, 0, size, base_ptr);
+#else
   mem->glbl.base
     = (LPTSTR) MapViewOfFileExNuma (mem->shm_fd, FILE_MAP_ALL_ACCESS, 0, 0,
 				    size, base_ptr, NUMA_NO_PREFERRED_NODE);
+#endif
   if (mem->glbl.base == NULL)
     {
       perror ("MapViewOfFile failed");
