diff --git a/bfd/bfdio.c b/bfd/bfdio.c
index 22c39a7b0cc..fd9f1c2d8c7 100644
--- a/bfd/bfdio.c
+++ b/bfd/bfdio.c
@@ -120,13 +120,8 @@ _bfd_real_fopen (const char *filename, const char *modes)
   /* PR 25713: Handle extra long path names possibly containing '..' and '.'.  */
    wchar_t **     lpFilePart = {NULL};
    const wchar_t  prefix[] = L"\\\\?\\";
-   const size_t   partPathLen = strlen (filename) + 1;
 #ifdef __MINGW32__
-#if !HAVE_DECL____LC_CODEPAGE_FUNC
-/* This prototype was added to locale.h in version 9.0 of MinGW-w64.  */
-   _CRTIMP unsigned int __cdecl ___lc_codepage_func (void);
-#endif
-   const unsigned int cp = ___lc_codepage_func ();
+   const unsigned int cp = CP_ACP;
 #else
    const unsigned int cp = CP_UTF8;
 #endif
@@ -141,8 +136,8 @@ _bfd_real_fopen (const char *filename, const char *modes)
    MultiByteToWideChar (cp, 0, filename, -1, partPath, partPathWSize);
 
    /* Convert any UNIX style path separators into the DOS i.e. backslash separator.  */
-   for (ix = 0; ix < partPathLen; ix++)
-     if (IS_UNIX_DIR_SEPARATOR(filename[ix]))
+   for (ix = 0; partPath[ix] != L'\0'; ix++)
+     if (partPath[ix] <= L'\x7f' && IS_UNIX_DIR_SEPARATOR ((char)partPath[ix]))
        partPath[ix] = '\\';
 
    /* Getting the full path from the provided partial path.
@@ -167,7 +162,7 @@ _bfd_real_fopen (const char *filename, const char *modes)
    /* It is non-standard for modes to exceed 16 characters.  */
    wchar_t    modesW[16];
 
-   MultiByteToWideChar (cp, 0, modes, -1, modesW, sizeof(modesW));
+   MultiByteToWideChar (cp, 0, modes, -1, modesW, sizeof(modesW) / sizeof(wchar_t));
 
    FILE *     file = _wfopen (fullPath, modesW);
    free (fullPath);
